File: bridging/OscillatorNode.cpp
#include "dart_api/dart_api.h"
#include "LabSound/LabSound.h"
#include "KeepNode.cpp"
#include "struct.h"
using namespace lab;

DART_EXPORT int createOscillatorNode(AudioContext* context) {
    auto sample = std::make_shared<OscillatorNode>(*context);
    return keepNode(sample);
}

DART_EXPORT int OscillatorNode_type(int nodeIndex) {
    return static_cast<int>(std::dynamic_pointer_cast<OscillatorNode>(audioNodes.find(nodeIndex)->second)->type());
}

DART_EXPORT void OscillatorNode_setType(int nodeIndex, int type) {
    std::dynamic_pointer_cast<OscillatorNode>(audioNodes.find(nodeIndex)->second)->setType(OscillatorType(type));
}

DART_EXPORT int OscillatorNode_amplitude(int nodeId) {
    return keepAudioParam(nodeId, 1, std::static_pointer_cast<OscillatorNode>(audioNodes.find(nodeId)->second)->amplitude());
}

DART_EXPORT int OscillatorNode_frequency(int nodeId) {
    return keepAudioParam(nodeId, 2, std::static_pointer_cast<OscillatorNode>(audioNodes.find(nodeId)->second)->frequency());
}

DART_EXPORT int OscillatorNode_detune(int nodeId) {
    return keepAudioParam(nodeId, 3, std::static_pointer_cast<OscillatorNode>(audioNodes.find(nodeId)->second)->detune());
}

DART_EXPORT int OscillatorNode_bias(int nodeId) {
    return keepAudioParam(nodeId, 4, std::static_pointer_cast<OscillatorNode>(audioNodes.find(nodeId)->second)->bias());
}

File: lab_sound_ffi/lib/src/lab_sound/oscillator_node.dart
import 'audio_context.dart';
import 'audio_param.dart';
import 'audio_scheduled_source_node.dart';
import 'lab_sound.dart';

enum OscillatorType {
  oscillatorNone,
  sine,
  fastSine,
  square,
  sawtooth,
  triangle,
  custom,
}

class OscillatorNode extends AudioScheduledSourceNode {

  late AudioParam amplitude;
  late AudioParam frequency;
  late AudioParam detune;
  late AudioParam bias;

  OscillatorNode(AudioContext ctx)
      : super(ctx, LabSound().createOscillatorNode(ctx.pointer)) {
    amplitude = AudioParam(this.ctx, this.nodeId, LabSound().OscillatorNode_amplitude(this.nodeId));
    frequency = AudioParam(this.ctx, this.nodeId, LabSound().OscillatorNode_frequency(this.nodeId));
    detune = AudioParam(this.ctx, this.nodeId, LabSound().OscillatorNode_detune(this.nodeId));
    bias = AudioParam(this.ctx, this.nodeId, LabSound().OscillatorNode_bias(this.nodeId));
  }

  OscillatorType get type => OscillatorType.values[LabSound().OscillatorNode_type(nodeId)];
  set type(OscillatorType type) => setType(type);

  setType(OscillatorType type) {
    LabSound().OscillatorNode_setType(nodeId, type.index);
  }

  @override
  dispose() {
    super.dispose();
  }
}
